test/functional/legacy/scroll_opt_spec.lua

```diff
@@ -589,14 +589,7 @@ describe('smoothscroll', function()
     -- Vim prefers to do so if we are scrolling a few lines only.
     exec("call setline(1, ['one', 'two', 'Line' .. (' with lots of text'->repeat(10)) .. ' end', 'four'])")
     feed('3Gztj')
-    screen:expect([[
-      <<<th lots of text with lots of text wit|
-      h lots of text with lots of text with lo|
-      ts of text with lots of text with lots o|
-      f text with lots of text end            |
-      ^four                                    |
-                                              |
-    ]])
+    screen:expect_unchanged()
```

Comment:
After changing the order of the tests before this, this now triggers an assert in screen.lua🤔:
```
test/functional/helpers.lua:244: test/functional/ui/screen.lua:656: _handle_grid_cursor_goto failed
  payload: { "grid_cursor_goto", { 1, -5, 0 } }
  error:   test/functional/ui/screen.lua:790: assertion failed!
```
https://github.com/neovim/neovim/blob/4e5061dba765df2a74ac4a8182f6e7fe21da125d/test/functional/ui/screen.lua#L790
What is the significance of that assert @bfredl? Can it be that it is harmless here?

Reply:
I think the problem is in `f_setline()`. A corrupt display/cursor position is visible after the `call setline()` step in that test. It is present in vim as well. i guess skipcol is not or not correctly computed after changing the buffer content through `setline()`.

Reply:
If you have a split above you'll see what's the problem.

Reply:
This fixes the problem, I'll submit a patch to vim.

```diff
@@ -272,6 +274,14 @@ set_buffer_lines(
                changed_bytes(lnum, 0);
                if (is_curbuf && lnum == curwin->w_cursor.lnum)
                    check_cursor_col();
+               // Reset "w_skipcol" if a currently visible line is replaced.
+               FOR_ALL_TAB_WINDOWS(tp, wp) {
+                   if (wp->w_buffer == buf
+                           && lnum < wp->w_botline
+                           && lnum >= wp->w_topline) {
+                       wp->w_skipcol = 0;
+                   }
+               }
                rettv->vval.v_number = 0;       // OK
            }
        }
```